plugins/tradeMetricsHandler.go

```diff
@@ -0,0 +1,42 @@
+package plugins
+
+import (
+	"github.com/stellar/kelp/model"
+)
+
+// TradeMetricsHandler tracks the number of trades
+type TradeMetricsHandler struct {
+	trades []model.Trade
+}
+
+// MakeTradeMetricsHandler is a factory method for the TradeMetricsHandler
+func MakeTradeMetricsHandler() *TradeMetricsHandler {
+	return &TradeMetricsHandler{
+		trades: []model.Trade{},
+	}
+}
+
+// Reset sets the handler's trade counter to zero.
+func (h *TradeMetricsHandler) Reset() {
+	h.trades = []model.Trade{}
+}
+
+// Read stores new trades internally.
+func (h *TradeMetricsHandler) Read(newTrades []model.Trade) {
+	for _, nt := range newTrades {
+		h.trades = append(h.trades, nt)
+	}
+}
+
+// NumTrades returns the number of trades.
+func (h *TradeMetricsHandler) NumTrades() int {
+	return len(h.trades)
+}
+
+// TotalBaseVolume returns the total base volume.
+func (h *TradeMetricsHandler) TotalBaseVolume() (total float64) {
```

Comment:
👍

Reply:
we should add additional metrics here -- maybe can return a map instead of individual metrics and call the function `ComputeMetrics` or something similar to indicate that we are (a) running calculations and (b) we will return the metrics result that we wanted to collect from this handler.

(Note: base units is `model.Trade.Amount` and quote units is `model.Trade.Amount * model.Trade.Price`)
- totalBaseVolume (buy and sell action are both +ve) -- buying $100 and selling $100 should equal $200
- totalQuoteVolume (buy and sell action are both +ve)
- netBaseVolume (buy action is +ve, sell is -ve) -- buying $100 and selling $100 should equal $0
- netQuoteVolume (buy action is -ve, sell action is +ve)
- numTrades
- avgTradeSizeBase (totalBaseVolume/numTrades)
- avgTradeSizeQuote (totalQuoteVolume/numTrades)
- avgTradePrice (simple average of the trade price)
- [vwap](https://en.wikipedia.org/wiki/Volume-weighted_average_price) (volume-weighted average of the trade price) -- I think this simplifies to `totalQuoteVolume/totalBaseVolume` but you'll have to confirm

anything else you can think of that would be important to capture for trades?

Reply:
PS: we will need to combine this with `Reset()` and make sure that we create a copy of the trades structure and then reset it, and then run our computations. This will ensure that there is a very small window in which we lose trades. We should also add a TODO NS at that point so we add proper locking around the `trades` field on the metrics handler, since we will face contention for it when writing from the `HandleFills` function which will be called from a separate thread.

Reply:
I think this is everything (including that VWAP formula) -- super helpful, thank you!! I've made these changes.